### PR TITLE
Optimized reads

### PR DESCRIPTION
**Pull request by listx**

### CODE REVIEW

File: lib/dockerregistry/inventory.go [hunk 1]
```diff
@@ -1434,19 +1439,71 @@ func mkPopulateRequestsForPromotionEdges(
...
+	mfests []Manifest,
+	readRepos bool,
+) map[PromotionEdge]interface{} {
 
 	edges := ToPromotionEdges(mfests)
+
+	if readRepos {
+		regs := getRegistriesToRead(edges)
+		for _, reg := range regs {
+			fmt.Println("reading this reg:", reg)
```
Comment (justinsb) [inline lib/dockerregistry/inventory.go]: Is this leftover debug code?  klog maybe?

File: lib/dockerregistry/inventory.go [hunk 2]
```diff
@@ -1434,19 +1439,71 @@ func mkPopulateRequestsForPromotionEdges(
...
+			// Do not read these registries recursively, because we already know
+			// exactly which repositories to read (getRegistriesToRead()).
+			false,
+			MkReadRepositoryCmdReal)
+	}
+
 	return sc.getPromotionCandidates(edges)
 }
 
+// getRegistriesToRead collects all unique Docker repositories we want to read
```
Comment (justinsb) [inline lib/dockerregistry/inventory.go]: Should this be called getRepositoriesToRead ?  Admittedly the difference gets me every time.

File: lib/dockerregistry/inventory.go [hunk 3]
```diff
@@ -1434,19 +1439,71 @@ func mkPopulateRequestsForPromotionEdges(
...
+// from. This way, we don't have to read the entire Docker registry, but only
+// those paths that we are thinking of modifying.
+func getRegistriesToRead(
+	edges map[PromotionEdge]interface{}) []RegistryContext {
+
+	rcs := make(map[RegistryContext]interface{})
+
+	// Save the src and dst endpoints as registries. We only care about the
+	// registry and image name, not the tag or digest; this is to collect all
+	// unique Docker repositories that we want to care about.
```
Comment (justinsb) [inline lib/dockerregistry/inventory.go]: Drop "want to" I think

**Comment by justinsb** [comment]
A few comments on the new commit, but it LGTM and I'll re-review after the rebase!

File: lib/dockerregistry/inventory.go (hunk 2 above)
Comment (listx) [reply lib/dockerregistry/inventory.go]: I feel the same --- but I will do a global s/registry/repository in the future (don't want to tackle that here...).

File: lib/dockerregistry/inventory.go (hunk 1 above)
Comment (listx) [reply lib/dockerregistry/inventory.go]: Oops, yes, I'll use klog.

**Pull request merged**